### PR TITLE
Anchor tags should not be used as buttons

### PR DESCRIPTION
**Pull request by Matrix278**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Screenshot(s)
image

## Description

Resolves: #104

## Checklist:

- [x] I have thoroughly read the [CONTRIBUTING](../CONTRIBUTING.md) guidelines.
- [x] I understand my pull request will be thoroughly reviewed at high detail.
- [x] I understand the work in my pull request will **only** be available in the next version release of Checka11y.css and not in the current version release.
- [x] I confirm the work in this pull request is valid according to my findings and is not something for anything personal.
- [x] I have updated the [README](../README.md) and/or [features.md](../features.md) where and if applicable (still put an `x` if you have considered this but thought there was nothing to add or modify).
- [x] I have added myself to the `contributors` section in `package.json` (still put an `x` if you have considered this but decided not to add yourself).
- [x] I have checked I have **not** committed any **accidental** files.
- [ ] I have tested all the main modern browsers (I.e. Chrome, Firefox, Edge, Safari - please leave this unchecked if there were any browsers listed you could not test and list them in the [help](#help) section with details why you couldn't test that browser)
- [x] I have run the automated tests and added new ones to cover new code.
- [ ] All new and existing a11y checks still work correctly (compare your local `test/index.html` to the `test/index.html` in the `master` branch).

## Help

Could not test in Safari because I do not have access to an Apple device

There is one issue with an existing feature `E0009`, where the invalid HTML elements are nested inside a button. As this new feature is checking for empty `href="#"`, it will trigger a new message here as well. How I could change this? Does here should be added check if <a> is not a child of a button or it could be done another way? Need some help with that, what I could do about this?
```
<button>
        <a href="#">Hello, I am a link inside a &lt;button&gt;</a>
</button>
```
image

### CODE REVIEW

File: checka11y-warnings.css [hunk 1]
```diff
@@ -42,7 +42,6 @@
   text-decoration: none;
   text-transform: initial;
   text-shadow: none;
-  content: "WARNING (W0009): Focusable -moz-element in a aria-hidden element." !important;
```
Comment (Matrix278) [inline checka11y-warnings.css:45]: Not sure why this got deleted. Will revert this

**Comment by jackdomleo7** [comment]
Hi @Matrix278, thank you for this PR! That's an interesting thing you have found. Are you saying the warning W0010 overrides the error if both apply? 🤔

File: codes.md [hunk 2]
```diff
@@ -117,4 +117,7 @@ A list of every Checka11y.css error & warning code with details on what the issu
...
 
 - ### W0009
-  The highlighted element is a focusable element that is nested within another element with `aria-hidden="true"`. This means the focusable element is inaccessible to assistive technologies. Either remove the `aria-hidden="true"`, or restructure the HTML so that the focusable element is not nested within the element with `aria-hidden="true"`. [Read more about this here](https://web.dev/aria-hidden-focus).
\ No newline at end of file
+  The highlighted element is a focusable element that is nested within another element with `aria-hidden="true"`. This means the focusable element is inaccessible to assistive technologies. Either remove the `aria-hidden="true"`, or restructure the HTML so that the focusable element is not nested within the element with `aria-hidden="true"`. [Read more about this here](https://web.dev/aria-hidden-focus).
+
+- ### W0010
```
Comment (jackdomleo7) [inline codes.md:122]: Really clear description. Just one minor change needed to the below sentence

"Those anchor elements ~with~ are often used to trigger a click event on the page."

**Comment by Matrix278** [comment]
> Hi @Matrix278, thank you for this PR! That's an interesting thing you have found. Are you saying the warning W0010 overrides the error if both apply? 🤔

Hi, yes it seems like it overrides it and showing the new warning there now, where actually should be error

File: package.json [hunk 3]
```diff
@@ -61,7 +62,7 @@
   "devDependencies": {
     "autoprefixer": "^10.2.5",
     "cypress": "^6.6.0",
-    "node-sass": "^5.0.0",
+    "node-sass": "^6.0.1",
```
Comment (jackdomleo7) [inline package.json:65]: Could you revert this change? I'm currently doing some project refactoring that also involves removing the use of `node-sass`

File: src/warnings/features/_links.scss [hunk 4]
```diff
@@ -4,3 +4,12 @@ a[target = "_blank"] {
     @include contentMessage(warning, "W0006", "Links that open in a new tab or window should be communicated to the user, ensure the user knows of this behavior.");
   }
 }
+
+/* W0010: Anchor tags should not be used as buttons. Links should redirect to a resource/page, if they don't they probably should be buttons. */
+a[href = "#"],
```
Comment (jackdomleo7) [inline src/warnings/features/_links.scss:9]: Nice CSS! Maybe could do with a minor refactor:

```scss
a {
  &[href = "#"],
  &[role = "button"],
  &[href ^= "javascript:"] {
    &::after {
      @include contentMessage(warning, "W0010", "Anchor tags should not be used as buttons. Links should redirect to a resource/page, if they don't they probably should be buttons.");
    }
  }
}
```

**Review by jackdomleo7 — CHANGES_REQUESTED** [review]
Thank you for this PR. Really good work. Just a few minor changes then we'd be good to go 🙂

**Comment by jackdomleo7** [comment]
That's interesting. I don't think any additional CSS is needed. After your PR is merged, I'll do some investigation before I publish a new version because I think I have an idea what it may be.

---

FYI - You can ignore the linting check failure, there is an open issue (#87) where I'm struggling to resolve this.

File: codes.md (hunk 2 above)
Comment (Matrix278) [reply codes.md:122]: Yep, sure. Edited this description several times, needed to reread it :smile:

File: src/warnings/features/_links.scss (hunk 4 above)
Comment (Matrix278) [reply src/warnings/features/_links.scss:9]: Aha thats nice, thanks for showing, didn't knew that

**Comment by Matrix278** [comment]
> That's interesting. I don't think any additional CSS is needed. After your PR is merged, I'll do some investigation before I publish a new version because I think I have an idea what it may be.
> 
> FYI - You can ignore the linting check failure, there is an open issue (#87) where I'm struggling to resolve this.

Sure, thanks for clarifying :+1:

**Comment by jackdomleo7** [comment]
Hi @Matrix278, we have recently merged #111. Would you be able to update your branch with master and where you have added your new feature as W0010, could you change it to W0011 as W0010 now already exists due to the recent merge 🙂

**Comment by Matrix278** [comment]
@jackdomleo7 ready for review again, fixed all your suggestions

**Review by jackdomleo7 — APPROVED** [review]
Nice one @Matrix278! Thank you.
